subsys/bluetooth/mesh/proxy_srv.c

```diff
@@ -585,29 +586,44 @@ static int gatt_proxy_advertise(struct bt_mesh_subnet *sub)
 		}
 	}
 
-	if (sub->node_id == BT_MESH_NODE_IDENTITY_RUNNING) {
-		uint32_t active = k_uptime_get_32() - sub->node_id_start;
+	for (int i = 0; i < subnet_count; i++) {
 
-		if (active < NODE_ID_TIMEOUT) {
-			remaining = NODE_ID_TIMEOUT - active;
-			BT_DBG("Node ID active for %u ms, %d ms remaining",
-			       active, remaining);
-			err = node_id_adv(sub, remaining);
-		} else {
-			bt_mesh_proxy_identity_stop(sub);
-			BT_DBG("Node ID stopped");
+		if (sub->node_id == BT_MESH_NODE_IDENTITY_RUNNING) {
+			uint32_t active = k_uptime_get_32() - sub->node_id_start;
+
+			if (active < NODE_ID_TIMEOUT) {
+				remaining = NODE_ID_TIMEOUT - active;
+				BT_DBG("Node ID active for %u ms, %d ms remaining",
+				       active, remaining);
+				err = node_id_adv(sub, remaining);
+				planned = true;
+			} else {
+				bt_mesh_proxy_identity_stop(sub);
+				BT_DBG("Node ID stopped");
+			}
 		}
-	}
 
-	if (sub->node_id == BT_MESH_NODE_IDENTITY_STOPPED) {
-		err = net_id_adv(sub, remaining);
-	}
+		/* Mesh Profile Specification v1.0.1, section 7.2.2.2.1
+		 * A node that does not support the Proxy feature or
+		 * has the GATT Proxy state disabled shall not advertise with Network ID.
+		 */
+		if (sub->node_id == BT_MESH_NODE_IDENTITY_STOPPED &&
+			bt_mesh_gatt_proxy_get() == BT_MESH_FEATURE_ENABLED) {
+			err = net_id_adv(sub, remaining);
+			planned = true;
+		}
 
-	BT_DBG("Advertising %d ms for net_idx 0x%04x", remaining, sub->net_idx);
+		beacon_sub = bt_mesh_subnet_next(beacon_sub);
```

Comment:
Will it be possible that `bt_mesh_subnet_next` returns the same subnet, but `planned` is false and `i` is less than `subnet_count`? Maybe use `next_sub()` instead? Also, `sub` value passed as an argument is ignored (see line 565).

Reply:
Nop, `bt_mesh_subnet_next` iterates through subnetworks that have initialized key as well as `bt_mesh_subnet_find` that is used to calculate subnetworks count. 
It is not correct to use `next_sub()` since it iterates only through subnetworks with `Node ID` in progress. That means if there are a couple of subnetworks without active `Node ID` they will not be shuffled here. 
You are right it is more correct to pass here `sub` instead of `beacon_sub`. It will work similarly except for the first cycle when the zeroes subnetwork is handled twice. Thanks.